src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -155,30 +161,53 @@ static async ValueTask Await(int i, ValueTask vt, List<object> toDispose)
                         ((IDisposable)disposable).Dispose();
                     }
                 }
+
+                scope.ClearState();
+            }
+        }
+
+        private IDictionary<ServiceCacheKey, object> ScopeDisposed()
+        {
+            ThrowHelper.ThrowObjectDisposedException();
+            return null;
+        }
+
+        private void ClearState()
+        {
+            // We lock here since ResolvedServices is always accessed in the scope lock, this means we'll never
+            // try to return to the pool while somebody is trying to access ResolvedServices.
+            lock (_scopeLock)
+            {
+                // Dispose the state, which will end up attempting to return the state pool.
+                // This will return false if the pool is full or if this state object is the root scope
+                if (_state.Return())
+                {
+                    _state = null;
+                }
             }
         }
 
         private List<object> BeginDispose()
         {
-            List<object> toDispose;
-            lock (_disposelock)
+            lock (_scopeLock)
             {
                 if (_disposed)
                 {
                     return null;
                 }
 
+                // We've transitioned to the disposed state, so future calls to
+                // CaptureDisposable will immediately dispose the object.
+                // No further changes to _state.Disposables, are allowed.
                 _disposed = true;
-                toDispose = _disposables;
-                _disposables = null;
+
+                return _state.Disposables;
 
                 // Not clearing ResolvedServices here because there might be a compilation running in background
                 // trying to get a cached singleton service instance and if it won't find
                 // it it will try to create a new one tripping the Debug.Assert in CaptureDisposable
                 // and leaking a Disposable object in Release mode
```

Comment:
Should we move this comment above the call to `_state.Return()` in `ClearState()`? At first I was going to point out we are clearing `ResolvedServices` now, but then I realized that we're not for the root scope.

Reply:
Or maybe `ScopePool.Return()` would now be a better place for this comment.